### DISCUSSION TITLE
Looking forward… v2.0

### DISCUSSION THREAD

**Discussion by bazmonk** [bazmonk/digigurdy-baz#46] [General]
V1.0 is coming up soon, time to start thinking about where to take it from here…

- [ ] Code improvement: In re-writing the code I created classes like a good boy for many of the gurdy functions, but a lot of the logic is still bare in the loop() and setup().  Some things appear in odd places in the code for the sake of being defined in the right order.  This can all be a little better.  I’d like to extend what the HurdyGurdy() class does to manage more of the playing logic.  Before it gets out of hand I wanna tighten it up some more.
- [ ] More strings: We’ve got two more channels.  I want them to be optional drones OR chanters.  I’ve done more playing around with scale tuning on bs-16i and I think they can be useful.
- [ ] Improved menu interface: Besides tuning, the menu could be more sophisticated.  Multiple save slots.  String on/off functions.  MIDI volume adjustment.  It could handle a lot more with some planning.
- [x] “Official” Teensy 4.1 support.  4.1 should be able to work as a drop-in replacement on DigiGurdies for the 3.5 they’re normally built with.  I’ve got one report that this code works on it, and I ordered my own unit to test.  This will not only give makers more flexibility, but users can more easily replace their units, and in the future the improved speed and power of the 4.1 may come in handy.

I’ll flesh this out more later, just getting thoughts down on “paper”.

**Comment by Gubbledenut** [comment]
Do pleaser allow an option to just select the original 5 tunings menu as well as the complex one, it's stillnpretty adaptable with the capo and octave keys

**Comment by bazmonk** [comment]
@Gubbledenut I probably won’t implement the original menu verbatim, but I can incorporate a “presets” screen that’s easy to access and provides the original tunings.

I was thinking of some number of presets, that would come pre-populated with the original tuning choices.  Users would have an option to overwrite them if they have a different one they want quickly available, and there would be an option to reset the slots to those original presets.  That should give the ease of the original interface (preset options will be quickly accessible with about the same number of key presses to get there), but the flexibility of the new one.

I’m also probably not going to include an option for the octave keys to act as octave keys again, since the transposing feature can accomplish the same thing (max range is +/- one octave) with far more flexibility, and the higher-octave tuning itself is also probably available in the tuning menu.

**Comment by bazmonk** [comment]
Another 2.0 thought: stash a magic number somewhere in the EEPROM, and use that for the code to determine if it has been cleared or not.  Code could look for that key, and if it's not there automatically clear the EEPROM for first-time use.

Would avoid users having to know about and go through that step to avoid problems with fresh installations.

**Comment by bazmonk** [comment]
More ideas: 

* On a real gurdy, for just a split second, the string's pitch spikes as the tangents make contact with the vibrating string, making that tell-tale twang that the digigurdy doesn't have.

I wonder if, with the synth's support, a very rapid spike in the pitch bend could give this effect.  It would require multi-threading, as each note would have to be tracking when it was initially pressed to adjust the pitch bend without interrupting the rest of play.

* In the create-new-tuning menu where you are offered strings that make sense in G/C or D/G, the first choice is which key you want to choose strings in.  For crazy people like me, I want a third option there: full-manual tuning.  That will let you choose each string (and the buzz note!) individually and set it to literally anything you want.  And it will do so without being in the way of users that don't want to mess with it.

Basically I saw a video of someone playing with G melody strings and a drone on A, and I realized I couldn't make that tuning on the digigurdy... at least I couldn't figure out a tuning/transpose/capo combo to do it.  I like the current menu for giving people lots of choices within reason, but I think full adjustment should be possible for when you know exactly what you want.

* Also in that full-manual tuning menu tree, I want an option to turn on or off the "melody" function on strings.  This effectively means that any string can be a melody or drone.  The code is already a few if-statements and a menu option away from being able to do this.

* Since the buzz note will be adjustable, which one is effectively the trompette also changes.  You'll be able to have a buzz note that doesn't even correspond with one of your strings if you want.  You'll be able to have four melody strings if you want.  

THIS is really my vision of what should be available in terms of advanced options: the full flexibility of this thing should be exposed to users that want it.  

The current presets and create-new-tuning tree will remain, so the entire "crazy" side of it will stay out of the way.
